### PR TITLE
 libnm: init, libproxy: use libnm instead of full networkmanager

### PR DESCRIPTION
**Pull request by SuperSandro2000**
###### Description of changes

@jtojnar do you have any comments about this?

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jtojnar** [comment]
I think this makes sense in theory but for Nix, we probably care about build closure, so it will probably need more work.

Do you have numbers on the runtime closure reduction?

Why are you using separate flag for plug-ins?

**Comment by SuperSandro2000** [comment]
> I think this makes sense in theory but for Nix, we probably care about build closure, so it will likely need more work.

this is only about runtime closure.

> Do you have numbers on the runtime closure reduction?

```
 ➜ nix path-info -Sh ./result*
/nix/store/0qv89sww54rylfbfpr61q5hwqrl80hay-networkmanager-1.42.6        468.6M
/nix/store/5qaad13xxnx7md81mmqfx4ib2q8gqsy9-networkmanager-1.42.6        203.3M
/nix/store/dahjw7k8qbvwi2b073wk1ni5vh38qpi6-networkmanager-1.42.6-man    201.1K
/nix/store/rmiadcs6syal240da2rzn6fhwfwbxkc4-networkmanager-1.42.6-man    201.1K
```

> Why are you using separate flag for plug-ins?

🤷🏼 removed it again

File: pkgs/tools/networking/networkmanager/default.nix [hunk 1]
```diff
@@ -145,15 +146,18 @@ stdenv.mkDerivation rec {
     ppp
     libndp
     curl
-    mobile-broadband-provider-info
+    jansson
+    dbus # used to get directory paths with pkg-config during configuration
+  ] ++ lib.optionals libnmOnly [
```
Comment (jtojnar) [inline pkgs/tools/networking/networkmanager/default.nix:151]: Should not this be negated?

**Comment by jtojnar** [comment]
> > I think this makes sense in theory but for Nix, we probably care about build closure, so it will likely need more work.
> 
> this is only about runtime closure.

Right. But I am questioning whether this change, which adds a lot of complexity to the expression, has any use case as is. Because I believe that Nix cares about build time closure of its dependencies. Please correct me if I am mistaken.

File: pkgs/tools/networking/networkmanager/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/tools/networking/networkmanager/default.nix:151]: oh, right, derp

**Comment by SuperSandro2000** [comment]
> Because I believe that Nix cares about build time closure of its dependencies. Please correct me if I am mistaken.

Also runtime closure, see https://github.com/NixOS/nix/issues/8648 . We don't need to add 200MB of extra dependencies to nix which aren't really needed and also other distros do this split.

**Comment by jtojnar** [comment]
Well, runtime closure improvements are useless if you cannot use the dependency due to build closure size.

**Comment by SuperSandro2000** [comment]
I am not sure which parts you exactly mean. Can you give an example?

As far as I understood the linked issue this should fix the mentioned problem. If I didn't understood it correctly please point me at the detail I missed.

**Comment by jtojnar** [comment]
The issue just talks about closure size. I understand that to refer to build closure.

File: pkgs/tools/networking/networkmanager/default.nix [hunk 2]
```diff
@@ -145,15 +146,18 @@ stdenv.mkDerivation rec {
...
     bluez5
     dnsmasq
-    modemmanager
-    readline
-    newt
     libsoup
-    jansson
-    dbus # used to get directory paths with pkg-config during configuration
+    mobile-broadband-provider-info
+    modemmanager
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/networkmanager/default.nix:160]: I moved most plugin dependencies out here, so build and runtime closure size should both be significant smaller now.

**Comment by jtojnar** [comment]
Maybe we should instead bump libproxy – the latest version does not appear to use libnm.

**Comment by SuperSandro2000** [comment]
did that in #272353

**Comment by bryango** [comment]
The networkmanager closure is just terrifyingly large, depending on openconnect, systemd, dnsmasq among other things. For people like me which uses nix as a package manager (not NixOS), these are all dead weight :sob: 

And it seems that the libproxy bump turns out to be non-trivial... Can we merge this first before that is sorted out? :pleading_face: Because other apps that depend on networkmanager can also benefit from switching to libnm.

**Comment by jtojnar** [comment]
Hopefully, libproxy will not take much longer.

**Comment by jtojnar** [comment]
No longer needed with https://github.com/NixOS/nixpkgs/pull/272353

**Pull request closed without merging**